crates/cli/src/lockfile.rs

```diff
@@ -0,0 +1,58 @@
+use anyhow::Context;
+use std::path::{Path, PathBuf};
+
+#[derive(Debug)]
+/// A file used as an exclusive lock on access to another file.
+///
+/// Constructing a `Lockfile` creates the `path` with [`std::fs::File::create_new`],
+/// a.k.a. `O_EXCL`, erroring if the file already exists.
+///
+/// Dropping a `Lockfile` deletes the `path`, releasing the lock.
+///
+/// Used to guarantee exclusive access to the system config file,
+/// in order to prevent racy concurrent modifications.
+pub struct Lockfile {
+    lock_path: PathBuf,
+}
+
+impl Lockfile {
+    /// Acquire an exclusive lock on the file `lock_path`.
+    pub fn acquire(lock_path: impl AsRef<Path>) -> anyhow::Result<Self> {
+        // Ensure the directory exists before attempting to create the lockfile.
+        let parent = lock_path.as_ref().parent().unwrap();
+        if parent != Path::new("") {
+            std::fs::create_dir_all(parent).with_context(|| {
+                format!(
+                    "Unable to create directory {parent:?} to build lockfile for {:?}",
+                    lock_path.as_ref(),
```

Comment:
note this does degrade our error messages slightly in some places, but I think they are still pretty explanatory?